core/src/main/scala/caliban/execution/Field.scala

```diff
@@ -186,47 +185,60 @@ object Field {
-                  val t = rootType.types.getOrElse(f.typeCondition.name, fieldType)
+                  val typeCondName      = f.typeCondition.name
+                  val t                 = rootType.types.getOrElse(typeCondName, fieldType)
+                  val subtypeNames0     = subtypeNames(typeCondName, rootType)
+                  val isSubsetCondition = subtypeNames0.getOrElse(Set.empty)
                   loop(
                     f.selectionSet,
                     t,
                     fragment = Some(Fragment(Some(name), resolvedDirectives)),
-                    targets = Some(Set(f.typeCondition.name)),
-                    condition = subtypeNames(f.typeCondition.name, rootType)
-                  ) -> Some(f.typeCondition.name)
+                    targets = Some(Set(typeCondName)),
+                    condition = subtypeNames0
+                  ).map {
+                    case t @ (_, Some(c)) if isSubsetCondition(c) => t
+                    case (f1, _)                                  => (f1, Some(typeCondName))
+                  }
                 }
               } else None
-              _fields.getOrElse(Nil -> None)
+              _fields.getOrElse(Nil)
             }
           )
-          fields.foreach(addField(_, condition))
+          fields.foreach((addField _).tupled)
         case InlineFragment(typeCondition, directives, selectionSet)    =>
           val resolvedDirectives = directives.map(resolveDirectiveVariables(variableValues, variableDefinitionsMap))
           if (checkDirectives(resolvedDirectives)) {
-            val t        = innerType.possibleTypes
-              .flatMap(_.find(_.name.exists(typeCondition.map(_.name).contains)))
-              .orElse(typeCondition.flatMap(typeName => rootType.types.get(typeName.name)))
+            val typeName          = typeCondition.map(_.name)
+            val t                 = innerType.possibleTypes
+              .flatMap(_.find(_.name.exists(typeName.contains)))
+              .orElse(typeName.flatMap(rootType.types.get))
               .getOrElse(fieldType)
-            val typeName = typeCondition.map(_.name)
+            val subtypeNames0     = typeName.flatMap(subtypeNames(_, rootType))
+            val isSubsetCondition = subtypeNames0.getOrElse(Set.empty)
             loop(
               selectionSet,
               t,
               fragment = Some(Fragment(None, resolvedDirectives)),
               targets = typeName.map(Set(_)),
-              condition = typeName.flatMap(subtypeNames(_, rootType))
-            ).foreach(addField(_, typeName))
+              condition = subtypeNames0
+            ).foreach { case (f, c) =>
+              if (c.exists(isSubsetCondition)) addField(f, c)
+              else addField(f, typeName)
+            }
           }
       }
-      map.values().asScala.toList
+      val builder = List.newBuilder[(Field, Option[String])]
+      map.forEach { case ((_, cond), field) => builder += ((field, cond)) }
```

Comment:
Is this zio-prelude's forEach? There's no `foreach`?

Reply:
No this is Java's `LinkedHashMap#forEach`

Reply:
Aha, was confused by the casing